File: packages/cucumber-universal-tests/features/steps/dataset_variant_steps.py
import json
from pathlib import Path

from behave import given, then, when


VARIANT_FIXTURE_PATH = (
    Path(__file__).resolve().parents[2]
    / "fixtures"
    / "historical.dataset-variants.json"
)


def load_variant_fixture():
    with VARIANT_FIXTURE_PATH.open("r", encoding="utf-8") as handle:
        return json.load(handle)


@given("multiple historical Sigillum data variants are available")
def load_historical_variants(context):
    context.variant_fixture = load_variant_fixture()
    context.historical_variants = context.variant_fixture["variants"]


@when("a variant is selected")
def select_historical_variant(context):
    variant_id = context.variant_fixture["defaultVariant"]
    context.selected_variant_id = variant_id
    context.selected_variant = context.historical_variants[variant_id]


@then("the selected variant identifies its source")
def assert_selected_variant_source(context):
    source = context.selected_variant.get("source")

    assert isinstance(source, dict)
    assert source.get("id")
    assert source.get("url")


@when("I inspect circumference division {division:d}")
def inspect_circumference_division(context, division):
    division_key = str(division)

    context.inspected_division = division
    context.variant_division_values = {
        variant_id: {
            "value": variant["circumferenceDivisions"][division_key],
            "source": variant["source"],
        }
        for variant_id, variant in context.historical_variants.items()
    }


@then("each differing value remains associated with its historical variant")
def assert_variant_value_associations(context):
    assert len(context.variant_division_values) >= 2

    for variant_id, reading in context.variant_division_values.items():
        assert variant_id
        assert reading["value"]
        assert reading["source"].get("id")
        assert reading["source"].get("url")


@then("the variants remain distinguishable")
def assert_variants_distinguishable(context):
    readings = context.variant_division_values

    assert len(set(readings.keys())) == len(readings)
    assert len({entry["value"] for entry in readings.values()}) > 1

File: packages/cucumber-universal-tests/features/step_definitions/dataset_variant_steps.rb
require "json"

VARIANT_FIXTURE_PATH = File.expand_path(
  "../../fixtures/historical.dataset-variants.json",
  __dir__
)

def load_variant_fixture
  JSON.parse(File.read(VARIANT_FIXTURE_PATH))
end

Given("multiple historical Sigillum data variants are available") do
  @variant_fixture = load_variant_fixture
  @historical_variants = @variant_fixture.fetch("variants")
end

When("a variant is selected") do
  @selected_variant_id = @variant_fixture.fetch("defaultVariant")
  @selected_variant = @historical_variants.fetch(@selected_variant_id)
end

Then("the selected variant identifies its source") do
  source = @selected_variant["source"]

  unless source.is_a?(Hash) &&
         !source["id"].to_s.empty? &&
         !source["url"].to_s.empty?
    raise "selected historical variant does not identify its source"
  end
end

When("I inspect circumference division {int}") do |division|
  division_key = division.to_s
  @inspected_division = division

  @variant_division_values = @historical_variants.to_h do |variant_id, variant|
    [
      variant_id,
      {
        "value" => variant.fetch("circumferenceDivisions").fetch(division_key),
        "source" => variant.fetch("source")
      }
    ]
  end
end

Then("each differing value remains associated with its historical variant") do
  raise "at least two variants are required" unless @variant_division_values.length >= 2

  @variant_division_values.each do |variant_id, reading|
    source = reading.fetch("source")

    if variant_id.to_s.empty? ||
       reading.fetch("value").to_s.empty? ||
       source.fetch("id").to_s.empty? ||
       source.fetch("url").to_s.empty?
      raise "variant reading lost its value or source association"
    end
  end
end

Then("the variants remain distinguishable") do
  variant_ids = @variant_division_values.keys
  values = @variant_division_values.values.map { |reading| reading.fetch("value") }

  raise "variant identifiers are not unique" unless variant_ids.uniq.length == variant_ids.length
  raise "variant values are not distinguishable" unless values.uniq.length > 1
end
